roachtest: quit-all-nodes failed

[(roachtest).quit-all-nodes failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1868732&tab=buildLog) on [master@6c5e395f16249ada2310864e66a11852930babe2](https://github.com/cockroachdb/cockroach/commits/6c5e395f16249ada2310864e66a11852930babe2):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200411-1868732/quit-all-nodes/run_1
	quit.go:402,quit.go:433,quit.go:142,context.go:135,quit.go:141,quit.go:433,test_runner.go:753: error with attached stack trace:
		    main.execCmdWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:426
		    main.(*cluster).RunWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2041
		    main.runQuit
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/quit.go:399
		    main.registerQuitAllNodes.func1.3
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/quit.go:433
		    main.(*quitTest).runWithTimeout.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/quit.go:142
		    github.com/cockroachdb/cockroach/pkg/util/contextutil.RunWithTimeout
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/util/contextutil/context.go:135
		    main.(*quitTest).runWithTimeout
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/quit.go:141
		    main.registerQuitAllNodes.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/quit.go:433
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:753
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - error with embedded safe details: %s
		    -- arg 1: <string>
		  - /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1868732-1586589776-18-n5cpu4:3 -- ./cockroach quit --insecure --logtostderr=INFO --port={pgport:3} --drain-wait=1h:
		  - signal: killed

```

<details><summary>More</summary><p>


Artifacts: [/quit-all-nodes](https://teamcity.cockroachdb.com/viewLog.html?buildId=1868732&tab=artifacts#/quit-all-nodes)
Related:
- #47328 roachtest: quit-all-nodes failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Aquit-all-nodes.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).quit-all-nodes failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1877809&tab=buildLog) on [master@8c928666de3f90cf17d81d70a3ee759c39e2ce27](https://github.com/cockroachdb/cockroach/commits/8c928666de3f90cf17d81d70a3ee759c39e2ce27):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/quit-all-nodes/run_1
	quit.go:402,quit.go:433,quit.go:142,context.go:135,quit.go:141,quit.go:433,test_runner.go:753: error with attached stack trace:
		    main.execCmdWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:426
		    main.(*cluster).RunWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2041
		    main.runQuit
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/quit.go:399
		    main.registerQuitAllNodes.func1.3
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/quit.go:433
		    main.(*quitTest).runWithTimeout.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/quit.go:142
		    github.com/cockroachdb/cockroach/pkg/util/contextutil.RunWithTimeout
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/util/contextutil/context.go:135
		    main.(*quitTest).runWithTimeout
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/quit.go:141
		    main.registerQuitAllNodes.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/quit.go:433
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:753
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - error with embedded safe details: %s
		    -- arg 1: <string>
		  - /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1877809-1586945944-18-n5cpu4:3 -- ./cockroach quit --insecure --logtostderr=INFO --port={pgport:3} --drain-wait=1h:
		  - signal: killed

```

<details><summary>More</summary><p>


Artifacts: [/quit-all-nodes](https://teamcity.cockroachdb.com/viewLog.html?buildId=1877809&tab=artifacts#/quit-all-nodes)
Related:
- #47328 roachtest: quit-all-nodes failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Aquit-all-nodes.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/cmd/roachtest/quit.go
+++ b/pkg/cmd/roachtest/quit.go
@@ -425,19 +425,25 @@ func registerQuitAllNodes(r *testRegistry) {
 			// everywhere for system ranges.
 			q.waitForUpReplication(ctx)
 
-			// Shut three nodes down gracefully with a very long wait (longer
+			// Shut one nodes down gracefully with a very long wait (longer
 			// than the test timeout). This is guaranteed to work - we still
 			// have quorum at that point.
 			q.runWithTimeout(ctx, func(ctx context.Context) { _ = runQuit(ctx, q.t, q.c, 5, "--drain-wait=1h") })
-			q.runWithTimeout(ctx, func(ctx context.Context) { _ = runQuit(ctx, q.t, q.c, 4, "--drain-wait=1h") })
-			q.runWithTimeout(ctx, func(ctx context.Context) { _ = runQuit(ctx, q.t, q.c, 3, "--drain-wait=1h") })
 
-			// Now shut down the remaining 3 nodes less gracefully, with a
-			// short wait. The drain should be stuck server-side, because
-			// the liveness record doesn't have quorum. So we're expecting
-			// quit to stop waiting due to the lower --drain-wait, but still
-			// issue a hard shutdown afterwards and then succeed. We can
-			// also expect the node to exit successfully.
+			// Now shut down the remaining 4 nodes less gracefully, with a
+			// short wait.
+
+			// For the next two nodes, we may or may not observe that
+			// the graceful shutdown succeed. It may succeed if every
+			// range has enough quorum on the last 2 nodes (shut down later below).
+			// It may fail if some ranges have a quorum composed of n3, n4, n5.
+			// See: https://github.com/cockroachdb/cockroach/issues/48339
+			q.runWithTimeout(ctx, func(ctx context.Context) { _ = runQuit(ctx, q.t, q.c, 4, "--drain-wait=4s") })
+			q.runWithTimeout(ctx, func(ctx context.Context) { _ = runQuit(ctx, q.t, q.c, 3, "--drain-wait=4s") })
+
+			// For the lat two nodes, we are always under quorum. In this
+			// case we can expect `quit` to always report a hard shutdown
+			// was required.
 			q.runWithTimeout(ctx, func(ctx context.Context) { expectHardShutdown(ctx, q.t, runQuit(ctx, q.t, q.c, 2, "--drain-wait=4s")) })
 			q.runWithTimeout(ctx, func(ctx context.Context) { expectHardShutdown(ctx, q.t, runQuit(ctx, q.t, q.c, 1, "--drain-wait=4s")) })
 

